tests/test_cms_menus.py

```diff
@@ -190,3 +247,446 @@ def test_get_roots_with_versioning_disabled(self):
         self.assertListEqual(
             list(roots), [menucontent_1.root, menucontent_2.root, menucontent_3.root]
         )
+
+
+class SoftrootTests(CMSTestCase):
+    """
+       Tree in fixture :
+               root
+                   aaa
+                       aaa1
+                           ccc
+                               ddd
+                       aaa2
+                   bbb
+       In the fixture, all pages are "in_navigation", "published" and
+       NOT-"soft_root".
+       What is a soft root?
+           A soft root is a page that acts as the root for a menu navigation tree.
+
+        Typically, this will be a page that is the root of a significant new
+        section on your site.
+
+        When the soft root feature is enabled, the navigation menu for any page
+        will start at the nearest soft root, rather than at the real root of
+        the site’s page hierarchy.
+
+        This feature is useful when your site has deep page hierarchies (and
+        therefore multiple levels in its navigation trees). In such a case, you
+        usually don’t want to present site visitors with deep menus of nested
+        items.
+       """
+    def setUp(self):
+        self.language = 'en'
+        self.client.force_login(self.get_superuser())
+
+    def assertTreeQuality(self, a, b, *attrs):
+        """
+        Checks that the node-lists a and b are the same for attrs.
+        This is recursive over the tree
+        """
+        msg = '%r != %r with %r, %r' % (len(a), len(b), a, b)
+        self.assertEqual(len(a), len(b), msg)
+        for n1, n2 in zip(a, b):
+            for attr in attrs:
+                a1 = getattr(n1, attr)
+                a2 = getattr(n2, attr)
+                msg = '%r != %r with %r, %r (%s)' % (a1, a2, n1, n2, attr)
+                self.assertEqual(a1, a2, msg)
+            self.assertTreeQuality(n1.children, n2.children, 'level', 'title')
```

Comment:
This should be: self.assertTreeQuality(n1.children, n2.children, *attrs), that will send the attrs back in.